### PR TITLE
ARROW-13613: [C++] Add decimal support to (hash) sum/mean/product

### PR DESCRIPTION
**Pull request by lidavidm**

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/aggregate_basic.cc [hunk 1]
```diff
@@ -187,31 +188,32 @@ struct ProductImpl : public ScalarAggregator {
...
     } else {
-      out->value = MakeScalar(this->product);
+      out->value = std::make_shared<OutputType>(this->product, out_type);
     }
     return Status::OK();
   }
 
-  size_t count = 0;
-  typename AccType::c_type product = 1;
+  std::shared_ptr<DataType> out_type;
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/aggregate_basic.cc:204]: Nit: const reference?

File: cpp/src/arrow/compute/kernels/aggregate_test.cc [hunk 2]
```diff
@@ -556,6 +641,98 @@ TYPED_TEST(TestNumericProductKernel, ScalarAggregateOptions) {
...
+  for (const auto& ty : {decimal128(3, 2), decimal256(3, 2)}) {
+    Datum null = ScalarFromJSON(ty, R"(null)");
+
+    EXPECT_THAT(Product(ArrayFromJSON(ty, R"([])")), ResultWith(null));
+    EXPECT_THAT(Product(ArrayFromJSON(ty, R"([null])")), ResultWith(null));
+    EXPECT_THAT(
+        Product(ArrayFromJSON(ty, R"(["0.00", "1.00", "2.00", "3.00", "4.00", "5.00"])")),
+        ResultWith(ScalarFromJSON(ty, R"("0.00")")));
+    Datum chunks =
+        ChunkedArrayFromJSON(ty, {R"(["1.00", "2.00", "3.00", "4.00", "5.00"])"});
```
Comment (cyb70289) [inline cpp/src/arrow/compute/kernels/aggregate_test.cc:654]: Nit: add some negative values?

**Review by cyb70289 — APPROVED** [review]

File: cpp/src/arrow/compute/kernels/aggregate_test.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/aggregate_test.cc:654]: Good pointed, updated the tests.

File: cpp/src/arrow/compute/kernels/aggregate_basic.cc (hunk 1 above)
Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/aggregate_basic.cc:204]: I don't think we want this as the reference may be invalidated.

**Comment by pitrou** [comment]
The changes here seem to include some of the changes in #10942?

**Comment by lidavidm** [comment]
> The changes here seem to include some of the changes in #10942?

I tried to reconcile them a little bit ahead of time to make my life easier since they will conflict on merge.

**Comment by lidavidm** [comment]
I'm going to move this to draft since I'd rather we get #10942 in first; there's going to be a lot of rebasing work I fear.

**Comment by pitrou** [comment]
Do you want to undraft this now?

**Comment by lidavidm** [comment]
Sorry, yes, let me rebase and undraft. Thanks for the reminder.

**Review by pitrou — APPROVED** [review]
+1, thanks for the update

**Pull request closed without merging**